designs/rerun_cli_commands.md

```diff
@@ -0,0 +1,85 @@
+
+```go
+if commandContext.OriginalError == nil {
+    // add task to completedTasks in checkpoint file
+}
+```
+
+When the user re-runs the command, we will check if the task exists as a completed task in the checkpoint file.
+If so, we will restore any necessary information from the checkpoint file and set the next task.
+
+```go
+for task != nil {
+    if _, ok := checkpointInfo.CompletedTasks[task.Name()]; ok {
+        // Restore any necessary information from the checkpoint file & set task to the next task
+        continue
+    }
+    logger.V(4).Info("Task start", "task_name", task.Name())
+    ...
+    ...
+}
+```
+
+### Other notable changes:
+* If the CLI fails to install CAPI objects to the bootstrap cluster, it deletes the bootstrap cluster.
+  This will obviously cause problems if we were to rerun the command at that point.
+  Therefore, in the `Restore()` method for the `CreateBootstrapClusterTask`, we will re-create the bootstrap cluster based on the cluster information in the `bootstrap-cluster-init` key from the checkpoint file if the bootstrap cluster does not exist.
+  An alternative to this would be to not delete the bootstrap cluster if `InstallCAPITask` fails & go straight to `CollectDiagnosticsTask`.
+
+
+* If the CLI fails when moving CAPI management, things start to get a bit tricky.
+  We could move CAPI management back to the source cluster if it fails, but this operation can be risky & potentially leave the cluster in an undesirable state.
+  Therefore, we will move this out of the scope of this initial MVP.
+  A deeper dive would be required to decide on a better way to roll back those changes if moving CAPI fails.
+  An idea for a future improvement is to backup the CAPI objects associated with the move using the `clusterctl backup` command. See [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/cmd/backup.go).
+  For now, if the cluster fails at this point (which does not happen terribly often), it will still fail, and we will handle that case in the future.
+
+
+* The only task that we should always run is the `setup-and-validate` task.
+  Skipping this task would leave a blind spot for misconfigurations from the user input.
+  However, if the CLI fails while CAPI management is temporarily in the bootstrap cluster, the preflight validations would fail because those CAPI objects are no longer present on the workload cluster.
+  To account for this, we could add a boolean flag to the command context to check whether or not CAPI management has been moved.
+  This flag can be set in the `Restore()` methods for the tasks where we are moving CAPI management.
+  If CAPI management has been moved from the workload cluster, we should skip the validation that checks if those objects are present on the cluster.
+
+### Future Work
+
+* Rolling back changes if moving CAPI management fails
```

Comment:
The ideal scenario is not having to rely on a checkpoint file and reading the state of the objects in the clusters to determine what phase we are in, similar to a reconciler. That might be a future work item too if we can remove the reliance of a checkpoint file.